File: ports/go/core/hnsw.go
package core

import (
	"container/heap"
	"math"
	"math/rand"
	"sort"
	"sync"
)

// A hand-rolled HNSW (Hierarchical Navigable Small World) index — a real approximate
// nearest-neighbour graph that runs in-process, so the cores don't need an external vector
// database. Implements Malkov & Yashunin (2016): a multi-layer navigable small-world graph
// with greedy descent through the upper layers and an ef-bounded best-first search at
// layer 0. Distances are 1 - cosine (smaller = nearer); scores returned are cosine
// similarity. The level-assignment RNG is seeded so recall is reproducible. Mirrors the
// Python/Java cores and the Flink in-JVM FlinkStateHnswVectorMemory.

// HnswIndex is an in-memory HNSW graph.
type HnswIndex struct {
	mu             sync.Mutex
	m              int
	m0             int
	efConstruction int
	efSearch       int
	ml             float64
	rng            *rand.Rand

	vecs  map[string][]float64
	text  map[string]string
	level map[string]int
	// graph[layer][nodeID] -> neighbour ids
	graph []map[string][]string
	entry string
	top   int
	count int
}

// NewHnswIndex builds an index. m neighbours per node per layer (2*m at layer 0),
// efConstruction candidate width while inserting, efSearch while querying. seed makes
// level assignment reproducible. Zero values fall back to sensible defaults.
func NewHnswIndex(m, efConstruction, efSearch int, seed int64) *HnswIndex {
	if m < 2 {
		m = 16
	}
	if efConstruction <= 0 {
		efConstruction = 200
	}
	if efSearch <= 0 {
		efSearch = 50
	}
	return &HnswIndex{
		m:              m,
		m0:             2 * m,
		efConstruction: efConstruction,
		efSearch:       efSearch,
		ml:             1.0 / math.Log(float64(m)),
		rng:            rand.New(rand.NewSource(seed)),
		vecs:           map[string][]float64{},
		text:           map[string]string{},
		level:          map[string]int{},
		entry:          "",
		top:            -1,
	}
}

func (h *HnswIndex) distance(a []float64, id string) float64 {
	return 1.0 - Cosine(a, h.vecs[id])
}

func (h *HnswIndex) neighbors(id string, layer int) []string {
	if layer >= len(h.graph) {
		return nil
	}
	return h.graph[layer][id]
}

func (h *HnswIndex) ensureLayers(level int) {
	for len(h.graph) <= level {
		h.graph = append(h.graph, map[string][]string{})
	}
}

func (h *HnswIndex) randomLevel() int {
	r := h.rng.Float64()
	for r <= 0.0 {
		r = h.rng.Float64()
	}
	return int(-math.Log(r) * h.ml)
}

// distItem is a (distance, id) pair used in the search heaps.
type distItem struct {
	dist float64
	id   string
}

// minHeap pops the nearest (smallest distance) first.
type minHeap []distItem

func (m minHeap) Len() int           { return len(m) }
func (m minHeap) Less(i, j int) bool { return m[i].dist < m[j].dist }
func (m minHeap) Swap(i, j int)      { m[i], m[j] = m[j], m[i] }
func (m *minHeap) Push(x any)        { *m = append(*m, x.(distItem)) }
func (m *minHeap) Pop() any          { old := *m; n := len(old); it := old[n-1]; *m = old[:n-1]; return it }

// maxHeap pops the farthest (largest distance) first.
type maxHeap []distItem

func (m maxHeap) Len() int           { return len(m) }
func (m maxHeap) Less(i, j int) bool { return m[i].dist > m[j].dist }
func (m maxHeap) Swap(i, j int)      { m[i], m[j] = m[j], m[i] }
func (m *maxHeap) Push(x any)        { *m = append(*m, x.(distItem)) }
func (m *maxHeap) Pop() any          { old := *m; n := len(old); it := old[n-1]; *m = old[:n-1]; return it }

// searchLayer does a best-first search within one layer, returning up to ef (distance, id)
// pairs sorted by ascending distance.
func (h *HnswIndex) searchLayer(query []float64, entryPoints []string, ef, layer int) []distItem {
	visited := make(map[string]bool, ef*2)
	candidates := &minHeap{}
	results := &maxHeap{}
	heap.Init(candidates)
	heap.Init(results)
	for _, ep := range entryPoints {
		d := h.distance(query, ep)
		visited[ep] = true
		heap.Push(candidates, distItem{d, ep})
		heap.Push(results, distItem{d, ep})
	}
	for candidates.Len() > 0 {
		c := heap.Pop(candidates).(distItem)
		worst := (*results)[0].dist
		if c.dist > worst && results.Len() >= ef {
			break
		}
		for _, e := range h.neighbors(c.id, layer) {
			if visited[e] {
				continue
			}
			visited[e] = true
			d := h.distance(query, e)
			worst = (*results)[0].dist
			if results.Len() < ef || d < worst {
				heap.Push(candidates, distItem{d, e})
				heap.Push(results, distItem{d, e})
				if results.Len() > ef {
					heap.Pop(results)
				}
			}
		}
	}
	out := make([]distItem, results.Len())
	copy(out, *results)
	sort.Slice(out, func(i, j int) bool { return out[i].dist < out[j].dist })
	return out
}

func (h *HnswIndex) selectNeighbors(candidates []distItem, m int) []string {
	sorted := make([]distItem, len(candidates))
	copy(sorted, candidates)
	sort.Slice(sorted, func(i, j int) bool { return sorted[i].dist < sorted[j].dist })
	if len(sorted) > m {
		sorted = sorted[:m]
	}
	ids := make([]string, len(sorted))
	for i, it := range sorted {
		ids[i] = it.id
	}
	return ids
}

func (h *HnswIndex) prune(id string, layer int) {
	mMax := h.m
	if layer == 0 {
		mMax = h.m0
	}
	neigh := h.graph[layer][id]
	if len(neigh) <= mMax {
		return
	}
	items := make([]distItem, len(neigh))
	for i, n := range neigh {
		items[i] = distItem{h.distance(h.vecs[id], n), n}
	}
	h.graph[layer][id] = h.selectNeighbors(items, mMax)
}

func (h *HnswIndex) greedy(query []float64, entry string, layer int) string {
	cur := entry
	curD := h.distance(query, cur)
	changed := true
	for changed {
		changed = false
		for _, e := range h.neighbors(cur, layer) {
			d := h.distance(query, e)
			if d < curD {
				cur, curD, changed = e, d, true
			}
		}
	}
	return cur
}

// Add inserts (or refreshes) a vector. Refreshing an existing id updates its vector/text
// and keeps existing links (approximate); typical KB usage inserts unique ids once.
func (h *HnswIndex) Add(id string, vector []float64, text string) {
	h.mu.Lock()
	defer h.mu.Unlock()

	_, update := h.vecs[id]
	v := make([]float64, len(vector))
	copy(v, vector)
	h.vecs[id] = v
	h.text[id] = text
	if update {
		return
	}
	h.count++

	level := h.randomLevel()
	h.level[id] = level
	h.ensureLayers(level)
	for lc := 0; lc <= level; lc++ {
		if _, ok := h.graph[lc][id]; !ok {
			h.graph[lc][id] = []string{}
		}
	}

	if h.entry == "" {
		h.entry = id
		h.top = level
		return
	}

	cur := h.entry
	for lc := h.top; lc > level; lc-- {
		cur = h.greedy(v, cur, lc)
	}
	start := level
	if h.top < start {
		start = h.top
	}
	for lc := start; lc >= 0; lc-- {
		found := h.searchLayer(v, []string{cur}, h.efConstruction, lc)
		m := h.m
		if lc == 0 {
			m = h.m0
		}
		for _, nb := range h.selectNeighbors(found, m) {
			h.graph[lc][id] = appendUnique(h.graph[lc][id], nb)
			h.graph[lc][nb] = appendUnique(h.graph[lc][nb], id)
			h.prune(nb, lc)
		}
		if len(found) > 0 {
			cur = found[0].id
		}
	}

	if level > h.top {
		h.top = level
		h.entry = id
	}
}

func appendUnique(list []string, v string) []string {
	for _, x := range list {
		if x == v {
			return list
		}
	}
	return append(list, v)
}

// Search returns the k nearest neighbours as Scored (score = cosine similarity).
func (h *HnswIndex) Search(query []float64, k int) []Scored {
	h.mu.Lock()
	defer h.mu.Unlock()
	if h.entry == "" {
		return nil
	}
	cur := h.entry
	for lc := h.top; lc > 0; lc-- {
		cur = h.greedy(query, cur, lc)
	}
	ef := h.efSearch
	if k > ef {
		ef = k
	}
	found := h.searchLayer(query, []string{cur}, ef, 0)
	if k > 0 && len(found) > k {
		found = found[:k]
	}
	out := make([]Scored, len(found))
	for i, it := range found {
		out[i] = Scored{ID: it.id, Score: 1.0 - it.dist, Text: h.text[it.id]}
	}
	return out
}

// Len reports the number of indexed vectors.
func (h *HnswIndex) Len() int {
	h.mu.Lock()
	defer h.mu.Unlock()
	return h.count
}

// HnswVectorStore adapts HnswIndex to the VectorStore SPI (in-process cold tier).
type HnswVectorStore struct {
	index *HnswIndex
}

// NewHnswVectorStore builds an in-process HNSW-backed vector store.
func NewHnswVectorStore(m, efConstruction, efSearch int, seed int64) *HnswVectorStore {
	return &HnswVectorStore{index: NewHnswIndex(m, efConstruction, efSearch, seed)}
}

func (s *HnswVectorStore) Upsert(docID string, embedding []float64, text string) {
	s.index.Add(docID, embedding, text)
}

func (s *HnswVectorStore) Search(query []float64, k int) []Scored { return s.index.Search(query, k) }

func (s *HnswVectorStore) ColdSearch() ColdSearch {
	return func(query []float64, k int) []Scored { return s.index.Search(query, k) }
}

File: ports/pyagentic/pyagentic/hnsw.py
"""A hand-rolled HNSW (Hierarchical Navigable Small World) index — a real approximate
nearest-neighbour graph that runs **in-process**, so the cores don't need an external
vector database. Implements Malkov & Yashunin (2016): a multi-layer navigable small-world
graph with greedy descent through the upper layers and an ``ef``-bounded best-first search
at layer 0.

Distances are ``1 - cosine`` (so smaller = nearer); scores returned to callers are the
cosine similarity. The level-assignment RNG is **seeded** so recall is reproducible in
tests. Mirrored by the Go and Java cores (same parameters and structure) and analogous to
the Flink in-JVM ``FlinkStateHnswVectorMemory``.
"""

from __future__ import annotations

import heapq
import math
import random
from typing import Dict, List, Optional, Tuple

from .retrieval import Scored, cosine


class HnswIndex:
    """In-memory HNSW graph. ``m`` neighbours per node per layer (``2*m`` at layer 0),
    ``ef_construction`` candidate width while inserting, ``ef_search`` while querying."""

    def __init__(self, m: int = 16, ef_construction: int = 200, ef_search: int = 50, seed: int = 42) -> None:
        if m < 2:
            raise ValueError("m must be >= 2")
        self.m = m
        self.m0 = 2 * m
        self.ef_construction = ef_construction
        self.ef_search = ef_search
        self._ml = 1.0 / math.log(m)
        self._rng = random.Random(seed)
        # node id -> (vector, text)
        self._vecs: Dict[str, List[float]] = {}
        self._text: Dict[str, str] = {}
        self._level: Dict[str, int] = {}
        # per-layer adjacency: _graph[layer][node_id] -> list of neighbour ids
        self._graph: List[Dict[str, List[str]]] = []
        self._entry: Optional[str] = None
        self._top = -1

    # ----- internals -------------------------------------------------------
    def _distance(self, a: List[float], node_id: str) -> float:
        return 1.0 - cosine(a, self._vecs[node_id])

    def _neighbors(self, node_id: str, layer: int) -> List[str]:
        if layer >= len(self._graph):
            return []
        return self._graph[layer].get(node_id, [])

    def _ensure_layers(self, level: int) -> None:
        while len(self._graph) <= level:
            self._graph.append({})

    def _random_level(self) -> int:
        # geometric distribution; rng in (0,1]
        r = self._rng.random()
        while r <= 0.0:
            r = self._rng.random()
        return int(-math.log(r) * self._ml)

    def _search_layer(self, query: List[float], entry_points: List[str], ef: int, layer: int) -> List[Tuple[float, str]]:
        """Best-first search within one layer; returns up to ``ef`` (distance, id) pairs."""
        visited = set(entry_points)
        # candidate min-heap by distance; results max-heap via negated distance
        candidates: List[Tuple[float, str]] = []
        results: List[Tuple[float, str]] = []
        for ep in entry_points:
            d = self._distance(query, ep)
            heapq.heappush(candidates, (d, ep))
            heapq.heappush(results, (-d, ep))
        while candidates:
            dist_c, c = heapq.heappop(candidates)
            worst = -results[0][0]
            if dist_c > worst and len(results) >= ef:
                break
            for e in self._neighbors(c, layer):
                if e in visited:
                    continue
                visited.add(e)
                d = self._distance(query, e)
                worst = -results[0][0]
                if len(results) < ef or d < worst:
                    heapq.heappush(candidates, (d, e))
                    heapq.heappush(results, (-d, e))
                    if len(results) > ef:
                        heapq.heappop(results)
        return sorted((( -nd), nid) for nd, nid in results)

    def _select_neighbors(self, candidates: List[Tuple[float, str]], m: int) -> List[str]:
        """Simple heuristic: keep the ``m`` closest candidates."""
        return [nid for _, nid in sorted(candidates)[:m]]

    def _prune(self, node_id: str, layer: int) -> None:
        m_max = self.m0 if layer == 0 else self.m
        neigh = self._graph[layer].get(node_id, [])
        if len(neigh) <= m_max:
            return
        scored = [(self._distance(self._vecs[node_id], n), n) for n in neigh]
        self._graph[layer][node_id] = [n for _, n in sorted(scored)[:m_max]]

    # ----- public API ------------------------------------------------------
    def add(self, node_id: str, vector: List[float], text: str = "") -> None:
        update = node_id in self._vecs
        self._vecs[node_id] = list(vector)
        self._text[node_id] = text
        if update:
            return  # vector refreshed; existing links kept (approximate)

        level = self._random_level()
        self._level[node_id] = level
        self._ensure_layers(level)
        for lc in range(level + 1):
            self._graph[lc].setdefault(node_id, [])

        if self._entry is None:
            self._entry = node_id
            self._top = level
            return

        cur = self._entry
        # greedy descent through layers above the new node's level
        for lc in range(self._top, level, -1):
            cur = self._greedy(self._vecs[node_id], cur, lc)

        for lc in range(min(level, self._top), -1, -1):
            found = self._search_layer(self._vecs[node_id], [cur], self.ef_construction, lc)
            m = self.m0 if lc == 0 else self.m
            selected = self._select_neighbors(found, m)
            for nb in selected:
                self._graph[lc].setdefault(node_id, [])
                self._graph[lc].setdefault(nb, [])
                if nb not in self._graph[lc][node_id]:
                    self._graph[lc][node_id].append(nb)
                if node_id not in self._graph[lc][nb]:
                    self._graph[lc][nb].append(node_id)
                self._prune(nb, lc)
            if found:
                cur = found[0][1]

        if level > self._top:
            self._top = level
            self._entry = node_id

    def _greedy(self, query: List[float], entry: str, layer: int) -> str:
        cur = entry
        cur_d = self._distance(query, cur)
        changed = True
        while changed:
            changed = False
            for e in self._neighbors(cur, layer):
                d = self._distance(query, e)
                if d < cur_d:
                    cur, cur_d, changed = e, d, True
        return cur

    def search(self, query: List[float], k: int) -> List[Scored]:
        if self._entry is None:
            return []
        cur = self._entry
        for lc in range(self._top, 0, -1):
            cur = self._greedy(query, cur, lc)
        ef = max(self.ef_search, k)
        found = self._search_layer(query, [cur], ef, 0)
        out = [Scored(nid, 1.0 - d, self._text.get(nid, "")) for d, nid in found[: max(1, k)]]
        return out

    def __len__(self) -> int:
        return len(self._vecs)
